proposals/list-patterns.md

```diff
@@ -73,16 +73,67 @@ Notes:
+A *list_pattern* is also compatible with any type that is *enumerable*.
 
-A *slice_pattern* is compatible with any type that is *countable* as well as *sliceable* - it has an accessible indexer that takes a `Range` argument or otherwise an accessible `Slice` method that takes two `int` arguments. If both are present, the former is preferred.
+A *slice_pattern* is compatible with any type that is *countable* as well as *sliceable* — it has an accessible indexer that takes a `Range` argument or otherwise an accessible `Slice` method that takes two `int` arguments. If both are present, the former is preferred.  
+A *slice_pattern* without a sub_pattern is also compatible with any type that is *enumerable*.
+
+```
+enumerable is { 1, 2, .. } // okay
+enumerable is { 1, 2, ..var x } // error
+```
 
 This set of rules is derived from the [***range indexer pattern***](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-8.0/ranges.md#implicit-index-support) but relaxed to ignore optional or `params` parameters, if any.
 
 > **Open question**: We should define the exact binding rules for any of these members and decide if we want to diverge from the range spec.
 
+#### Semantics on enumerable type
+
+If the input type is *enumerable* but not *countable*, then the *length_pattern* is checked on the number of elements obtained from enumerating the collection.
+
+If the input type is *enumerable* but not *indexable*, then the *list_pattern* enumerates elements from the collection and checks them against the listed patterns:  
+Patterns at the start of the *list_pattern* — that are before the `..` *slice_pattern* if one is present, or all otherwise — are matched against the elements produced at the start of the enumeration.  
+If the collection does not produce enough elements to get a value corresponding to a starting pattern, the match fails. So the *constant_pattern* `3` in `{ 1, 2, 3, .. }` doesn't match when the collection has fewer than 3 elements.  
+Patterns at the end of the *list_pattern* (that are following the `..` *slice_pattern* if one is present) are matched against the elements produced at the end of the enumeration.  
+If the collection does not produce enough elements to get values corresponding to the ending patterns, the *splice_pattern* does not match. So the *splice_pattern* in `{ 1, .., 3 }` doesn't match when the collection has fewer than 2 elements.  
+A *list_pattern* without a *splice_pattern* only matches if the number of elements produced by complete enumeration and the number of patterns are equals. So `{ _, _, _ }` only matches when the collection produces exactly 3 elements.
+
+Note that those implicit checks for number of elements in the collection are unaffected by the collection type being *countable*. So `{ _, _, _ }` will not make use of `Length` or `Count` even if one is available.
```

Comment:
I think if the type is countable we can just use it e.g. if we have Length but not an indexer, we match  Length while enumerating for elements.

Reply:
My thinking was that if we're enumerating to the end anyways, we might as well use the enumerated count that we've accumulated for the check at the end of the list-pattern.
But it's true that we could omit tailing discard patterns. So `{ 1, _, _ }` could be "check first element, check Count == 3".

On the other hand, imagine that the `Count` property needs to enumerate again from the start.

I'll add an open issue.

Reply:
> "check first element, check Count == 3".

We always check Count first if it's available without enumerating, subsequently `MoveNext() is 1 && MoveNext() && MoveNext() && !MoveNext()` will be emitted to match elements which also tests for the length in itself (in case it wasn't available)

Reply:
If we're going to emit 4 `MoveNext` evaluations and enumerate completely, then there is no need to also call `Count`/`Length`. We don't need to check the count twice.

Reply:
We dont *need* to do it but if we can get count without enumerating we'll check it first to fail early ifit doesn't match. This is from last LDM.

Reply:
What I'm saying is that when there is a count check for the closing brace `}` of the list-pattern on an enumerable, that should use the enumerated count, since we'll have enumerated anyways. Similarly, the check that occurs to confirm that we had at least 3 elements before applying the `3` would use the enumeration.

I think that in your framing, all I'm saying is that before we check whether pattern `3` matches, we must have had 3 successful `MoveNext()` calls and that at the closing brace `}` we check that `!MoveNext()`. Those checks just rely on the enumerator, there was no need to call `Count` on a countable type.

> If the enumerable is actually backed by a concrete list type, we don't need to do any enumeration of the enumerable to check the length pattern.

Yes, for the length-pattern (such as `[<6]`) it is fine to introduce a third concept of count, a "try-count", or extend the concept of the `Count` API, but that's orthogonal to what I'm saying about the list-pattern. 
Note that the `{ 1, 2, 3, .., 10 }` does not have an implicit end-of-list-pattern count, since it has a `..`. But it does have implicit checks that we could get at least 4 elements before we apply the pattern `10`.

I hope that makes sense.
By the way, if you're already down the implementation path and feel there is a more natural way to spec this, feel free to update this doc. I'll merge later today so you can stack a PR.

Reply:
Sorry, one more point of clarification.
This (`Note that those implicit checks for number of elements in the collection are unaffected by the collection type being *countable*. So `{ _, _, _ }` will not make use of `Length` or `Count` even if one is available.`) is merely calling out a consequence of what is stated in the preceding paragraph:

> If the collection does not produce enough elements to get a value corresponding to a starting pattern, the match fails. [...]

> If the collection does not produce enough elements to get values corresponding to the ending patterns, the *slice_pattern* does not match. [...]

> A *list_pattern* without a *slice_pattern* only matches if the number of elements produced by complete enumeration and the number of patterns are equals. [...]